File: Src/java/cql-to-elm/src/main/java/org/cqframework/cql/cql2elm/model/invocation/DateTimeInvocation.java
package org.cqframework.cql.cql2elm.model.invocation;

import org.hl7.elm.r1.DateTime;
import org.hl7.elm.r1.Expression;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class DateTimeInvocation extends OperatorExpressionInvocation {
    public DateTimeInvocation(DateTime expression) {
        super(expression);
    }

    @Override
    public Iterable<Expression> getOperands() {
        DateTime dt = (DateTime) expression;
        List<Expression> opList = Arrays.asList(dt.getYear(), dt.getMonth(), dt.getDay(), dt.getHour(), dt.getMinute(), dt.getSecond(), dt.getMillisecond(), dt.getTimezoneOffset());
        // If the last expression is null, we should trim this down
        int i;
        for (i = 7; i > 0 && opList.get(i) == null; i--);
        return opList.subList(0, i + 1);
    }

    @Override
    public void setOperands(Iterable<Expression> operands) {
        ArrayList<Expression> opList = new ArrayList<>();
        for (Expression operand : operands) {
            opList.add(operand);
        }
        setDateTimeFieldsFromOperands((DateTime) expression, opList);
    }

    public static void setDateTimeFieldsFromOperands(DateTime dt, List<Expression> operands) {
        if (operands.isEmpty() || operands.size() > 8) {
            throw new IllegalArgumentException(
                    "Could not resolve call to system operator DateTime.  Expected 1 - 8 arguments.");
        }
        dt.setYear(operands.get(0));
        if (operands.size() > 1) {
            dt.setMonth(operands.get(1));
        }
        if (operands.size() > 2) {
            dt.setDay(operands.get(2));
        }
        if (operands.size() > 3) {
            dt.setHour(operands.get(3));
        }
        if (operands.size() > 4) {
            dt.setMinute(operands.get(4));
        }
        if (operands.size() > 5) {
            dt.setSecond(operands.get(5));
        }
        if (operands.size() > 6) {
            dt.setMillisecond(operands.get(6));
        }
        if (operands.size() > 7) {
            dt.setTimezoneOffset(operands.get(7));
        }
    }
}

File: Src/java/cql-to-elm/src/commonMain/kotlin/org/cqframework/cql/cql2elm/model/invocation/DateTimeInvocation.kt
package org.cqframework.cql.cql2elm.model.invocation

import kotlin.jvm.JvmStatic
import org.hl7.elm.r1.DateTime
import org.hl7.elm.r1.Expression

class DateTimeInvocation(expression: DateTime) :
    OperatorExpressionInvocation<DateTime>(expression) {
    override var operands: List<Expression>
        get() =
            listOfNotNull(
                expression.year,
                expression.month,
                expression.day,
                expression.hour,
                expression.minute,
                expression.second,
                expression.millisecond,
                expression.timezoneOffset
            )
        set(operands) {
            setDateTimeFieldsFromOperands(expression, operands)
        }

    @Suppress("MagicNumber")
    companion object {
        @JvmStatic
        fun setDateTimeFieldsFromOperands(dt: DateTime, operands: List<Expression>) {
            require(operands.size in 1..8) { "DateTime operator requires one to eight operands." }

            dt.year = operands[0]
            if (operands.size > 1) {
                dt.month = operands[1]
            }
            if (operands.size > 2) {
                dt.day = operands[2]
            }
            if (operands.size > 3) {
                dt.hour = operands[3]
            }
            if (operands.size > 4) {
                dt.minute = operands[4]
            }
            if (operands.size > 5) {
                dt.second = operands[5]
            }
            if (operands.size > 6) {
                dt.millisecond = operands[6]
            }
            if (operands.size > 7) {
                dt.timezoneOffset = operands[7]
            }
        }
    }
}
